rmw_dds_common/src/qos.cpp

```diff
@@ -0,0 +1,317 @@
+  }
+
+  const rmw_time_t & pub_lease = publisher_qos.liveliness_lease_duration;
+  const rmw_time_t & sub_lease = subscription_qos.liveliness_lease_duration;
+  const rmw_time_t lease_default = RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT;
+
+  // No lease duration for publisher and lease duration for subscription
+  if (pub_lease == lease_default && sub_lease != lease_default) {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    rmw_ret_t append_ret = _append_to_buffer(
+      reason,
+      reason_size,
+      "ERROR: Subscription has a liveliness lease duration, but publisher does not;");
+    if (RMW_RET_OK != append_ret) {
+      return append_ret;
+    }
+  }
+
+  // Subscription lease duration is less than publisher lease duration
+  if (pub_lease != lease_default && sub_lease != lease_default) {
+    if (sub_lease < pub_lease) {
+      *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+      rmw_ret_t append_ret = _append_to_buffer(
+        reason,
+        reason_size,
+        "ERROR: Subscription liveliness lease duration is less than publisher;");
+      if (RMW_RET_OK != append_ret) {
+        return append_ret;
+      }
+    }
+  }
+
+  // Only check for warnings if there are no errors
+  if (RMW_QOS_COMPATIBILITY_OK == *compatibility) {
+    // We don't know the policy if the value is "system default" or "unknown"
+    const bool pub_reliability_unknown =
+      publisher_qos.reliability == RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT ||
+      publisher_qos.reliability == RMW_QOS_POLICY_RELIABILITY_UNKNOWN;
+    const bool sub_reliability_unknown =
+      subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_SYSTEM_DEFAULT ||
+      subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_UNKNOWN;
+    const bool pub_durability_unknown =
+      publisher_qos.durability == RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT ||
+      publisher_qos.durability == RMW_QOS_POLICY_DURABILITY_UNKNOWN;
+    const bool sub_durability_unknown =
+      subscription_qos.durability == RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT ||
+      subscription_qos.durability == RMW_QOS_POLICY_DURABILITY_UNKNOWN;
+    const bool pub_liveliness_unknown =
+      publisher_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT ||
+      publisher_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_UNKNOWN;
+    const bool sub_liveliness_unknown =
+      subscription_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT ||
+      subscription_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_UNKNOWN;
+
+    // Reliability for publisher is unknown and subscription is reliable
+    if (pub_reliability_unknown &&
+      subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE)
+    {
+      *compatibility = RMW_QOS_COMPATIBILITY_WARNING;
+      rmw_ret_t append_ret = _append_to_buffer(
+        reason,
+        reason_size,
+        "WARNING: Reliable subscription, but publisher is %s;",
+        rmw_qos_reliability_policy_to_str(publisher_qos.reliability));
```

Comment:
I was writing some more  tests and discovered the "*to_str" functions return `NULL` if the policy is unknown... I need to refactor this.

Reply:
Fixed in 6a67320